### PR TITLE
interfaces/builtin: rework for avahi interface

### PR DESCRIPTION
**Pull request by kubiko**
rework for avahi-observe
- removing classic only restriction
- adding slot declaration which can be now implemented by application snaps.
Adding new avahi-control for control level of capabilities

### CODE REVIEW

**Comment by zyga** [comment]
Hey. Is there any forum discussion on the background of this change?

File: interfaces/builtin/dbus_test.go [hunk 1]
```diff
@@ -292,6 +292,7 @@ plugs:
 
 func (s *DbusInterfaceSuite) TestPermanentSlotAppArmorSession(c *C) {
 	apparmorSpec := &apparmor.Specification{}
+	release.OnClassic = true
```
Comment (zyga) [inline interfaces/builtin/dbus_test.go]: Please use `release.MockOnClassic`

File: interfaces/builtin/dbus_test.go [hunk 2]
```diff
@@ -341,6 +342,7 @@ func (s *DbusInterfaceSuite) TestPermanentSlotAppArmorSessionClassic(c *C) {
 
 func (s *DbusInterfaceSuite) TestPermanentSlotAppArmorSystem(c *C) {
 	apparmorSpec := &apparmor.Specification{}
+	release.OnClassic = true
```
Comment (zyga) [inline interfaces/builtin/dbus_test.go]: Please use `release.MockOnClassic`

File: interfaces/builtin/avahi_control.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
```
Comment (zyga) [inline interfaces/builtin/avahi_control.go]: 2017

File: interfaces/builtin/avahi_control.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    allow-installation:
+      slot-snap-type:
+        - app
+        - core
+    deny-auto-connection: true
+    deny-connection:
+      on-classic: false
+`
+
+const avahiControlSummary = `allows control over local domains, hostnames and services`
```
Comment (zyga) [inline interfaces/builtin/avahi_control.go]: The summary doesn't really say this is about mDNS and service discovery.

File: interfaces/builtin/avahi_control.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+       Summary: avahiControlSummary,
+       ImplicitOnClassic: true,
+       BaseDeclarationSlots: avahiControlBaseDeclarationSlots,
+   }
+}
+
+func (iface *avahiControlInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.Plug, plugAttrs map[string]interface{}, slot *interfaces.Slot, slotAttrs map[string]interface{}) error {
+    old := "###SLOT_SECURITY_TAGS###"
+    var new string
+    if release.OnClassic {
```
Comment (zyga) [inline interfaces/builtin/avahi_control.go]: Nice

File: interfaces/builtin/avahi_control.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    spec.AddSnippet(snippet)
+    return nil
+}
+
+func (iface *avahiControlInterface) DBusPermanentSlot(spec *dbus.Specification, slot *interfaces.Slot) error {
+    spec.AddSnippet(avahiObservePermanentSlotDBus)
+    return nil
+}
+
+func (iface *avahiControlInterface) SanitizePlug(plug *interfaces.Plug) error {
```
Comment (zyga) [inline interfaces/builtin/avahi_control.go]: Can you please `ensurePlugIfaceMatch` here?

File: interfaces/builtin/avahi_control.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    return nil
+}
+
+func (iface *avahiControlInterface) SanitizeSlot(slot *interfaces.Slot) error {
+    return nil
```
Comment (zyga) [inline interfaces/builtin/avahi_control.go]: Can you please use `ensureSlotIfaceMatch` here?

File: interfaces/builtin/avahi_control_test.go [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+	plugSnap := snaptest.MockInfo(c, avahiControlMockPlugSnapInfoYaml, nil)
+	s.plug = &interfaces.Plug{PlugInfo: plugSnap.Plugs["avahi-control"]}
+}
+
+func (s *AvahiControlInterfaceSuite) TestName(c *C) {
+	c.Assert(s.iface.Name(), Equals, "avahi-control")
+}
+
+// The label glob when all apps are bound to the avahi slot
+func (s *AvahiControlInterfaceSuite) TestConnectedPlugSnippetUsesSlotLabelAll(c *C) {
```
Comment (zyga) [inline interfaces/builtin/avahi_control_test.go]: Can you please use `snaptest.MockInfo` and extract those from the resulting `snap.Info` instead?

File: interfaces/builtin/avahi_control_test.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	apparmorSpec := &apparmor.Specification{}
+	err := apparmorSpec.AddConnectedPlug(s.iface, s.plug, nil, slot, nil)
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SecurityTags(), DeepEquals, []string{"snap.other.app2"})
+	c.Assert(apparmorSpec.SnippetForTag("snap.other.app2"), testutil.Contains, `peer=(label="snap.avahi-control.*"),`)
+}
+
+// The label glob when all apps are bound to the avahi slot
+func (s *AvahiControlInterfaceSuite) TestConnectedPlugSnippetUsesSlotLabelSome(c *C) {
+	app1 := &snap.AppInfo{Name: "app1"}
```
Comment (zyga) [inline interfaces/builtin/avahi_control_test.go]: Same here

File: interfaces/builtin/avahi_control_test.go [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+	apparmorSpec := &apparmor.Specification{}
+	err := apparmorSpec.AddConnectedPlug(s.iface, s.plug, nil, slot, nil)
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SecurityTags(), DeepEquals, []string{"snap.other.app2"})
+	c.Assert(apparmorSpec.SnippetForTag("snap.other.app2"), testutil.Contains, `peer=(label="snap.avahi-control.{app1,app2}"),`)
+}
+
+// The label glob when all apps are bound to the avahi slot
+func (s *AvahiControlInterfaceSuite) TestConnectedPlugSnippetUsesSlotLabelOne(c *C) {
+	app := &snap.AppInfo{Name: "app"}
```
Comment (zyga) [inline interfaces/builtin/avahi_control_test.go]: And here

File: interfaces/builtin/avahi_control_test.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+			Snap: &snap.Info{
+				SuggestedName: "avahi-control",
+				Apps:          map[string]*snap.AppInfo{"app": app},
+			},
+			Name:      "avahi-control",
+			Interface: "avahi-control",
+			Apps:      map[string]*snap.AppInfo{"app": app},
+		},
+	}
+	release.OnClassic = false
```
Comment (zyga) [inline interfaces/builtin/avahi_control_test.go]: Please use `release.MockOnClassic` here

File: interfaces/builtin/avahi_control_test.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+			Snap: &snap.Info{
+				SuggestedName: "avahi-control",
+				Apps:          map[string]*snap.AppInfo{"app1": app1, "app2": app2},
+			},
+			Name:      "avahi-control",
+			Interface: "avahi-control",
+			Apps:      map[string]*snap.AppInfo{"app1": app1, "app2": app2},
+		},
+	}
+	release.OnClassic = false
```
Comment (zyga) [inline interfaces/builtin/avahi_control_test.go]: Please use `release.MockOnClassic` here

**Comment by kubiko** [comment]
@zyga not really, something I hit when snapping various projects. Also avahi snap as it exists now is quite limited compare to what avahi delivers on classic system. So this is just evening behaviour between UC and Classic on observer part of the interface. And control is just to go all the way on avahi interface. @jdstrand suggested to split it into observe and control parts which sounds reasonable.

Additionally avahi is often used when building bridges to Apple ecosystem, so common thing in IOT. homebridge project is good example.

Happy to create forum post though

File: interfaces/builtin/avahi_control_test.go [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+			Snap: &snap.Info{
+				SuggestedName: "avahi-control",
+				Apps:          map[string]*snap.AppInfo{"app1": app1, "app2": app2, "app3": app3},
+			},
+			Name:      "avahi-control",
+			Interface: "avahi-control",
+			Apps:      map[string]*snap.AppInfo{"app1": app1, "app2": app2},
+		},
+	}
+	release.OnClassic = false
```
Comment (zyga) [inline interfaces/builtin/avahi_control_test.go]: Please use `release.MockOnClassic` here

File: interfaces/builtin/avahi_control_test.go [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+				Apps:          map[string]*snap.AppInfo{"app": app},
+			},
+			Name:      "avahi-control",
+			Interface: "avahi-control",
+			Apps:      map[string]*snap.AppInfo{"app": app},
+		},
+	}
+	release.OnClassic = false
+
+	apparmorSpec := &apparmor.Specification{}
```
Comment (zyga) [inline interfaces/builtin/avahi_control_test.go]: Just say `spec` here, it's shorter. If you need multiple split those to individual tests. This will also clean up the problem of mocking the right data and you can move that up to the top of the file / into the suite type.

File: interfaces/builtin/avahi_control_test.go [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+	err := apparmorSpec.AddConnectedPlug(s.iface, s.plug, nil, slot, nil)
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SecurityTags(), DeepEquals, []string{"snap.other.app2"})
+	c.Assert(apparmorSpec.SnippetForTag("snap.other.app2"), testutil.Contains, `peer=(label="snap.avahi-control.app"),`)
+}
+
+func (s *AvahiControlInterfaceSuite) TestConnectedPlugSnippedUsesUnconfinedLabelOnClassic(c *C) {
+	slot := &interfaces.Slot{}
+	release.OnClassic = true
```
Comment (zyga) [inline interfaces/builtin/avahi_control_test.go]: Please use `release.MockOnClassic` here

File: interfaces/builtin/avahi_observe.go [hunk 16]
```diff
@@ -59,75 +262,207 @@ dbus (send)
...
+    return nil
+}
+
+func (iface *avahiObserveInterface) DBusPermanentSlot(spec *dbus.Specification, slot *interfaces.Slot) error {
+    spec.AddSnippet(avahiObservePermanentSlotDBus)
+    return nil
+}
+
+func (iface *avahiObserveInterface) SanitizePlug(plug *interfaces.Plug) error {
+    return nil
```
Comment (zyga) [inline interfaces/builtin/avahi_observe.go]: Same as above (use one of the new helpers).

File: interfaces/builtin/avahi_observe.go [hunk 17]
```diff
@@ -59,75 +262,207 @@ dbus (send)
...
+}
+
+func (iface *avahiObserveInterface) SanitizeSlot(slot *interfaces.Slot) error {
+    return nil
```
Comment (zyga) [inline interfaces/builtin/avahi_observe.go]: Same as above (use one of the new helpers).

File: interfaces/builtin/avahi_observe_test.go [hunk 18]
```diff
@@ -40,59 +41,139 @@ var _ = Suite(&AvahiObserveInterfaceSuite{
 	iface: builtin.MustInterface("avahi-observe"),
 })
 
-func (s *AvahiObserveInterfaceSuite) SetUpTest(c *C) {
-	var mockPlugSnapInfoYaml = `name: other
+const avahiMockPlugSnapInfoYaml = `name: other
 version: 1.0
 apps:
- app:
+ app2:
```
Comment (zyga) [inline interfaces/builtin/avahi_observe_test.go]: Why is this change necessary?

File: interfaces/builtin/avahi_observe_test.go [hunk 19]
```diff
@@ -40,59 +41,139 @@ var _ = Suite(&AvahiObserveInterfaceSuite{
...
+			Snap: &snap.Info{
+				SuggestedName: "avahi-observe",
+				Apps:          map[string]*snap.AppInfo{"app1": app1, "app2": app2, "app3": app3},
+			},
+			Name:      "avahi-observe",
+			Interface: "avahi-observe",
+			Apps:      map[string]*snap.AppInfo{"app1": app1, "app2": app2},
+		},
+	}
+	release.OnClassic = false
```
Comment (zyga) [inline interfaces/builtin/avahi_observe_test.go]: Please use the right mocking helper here.

File: interfaces/builtin/avahi_observe_test.go [hunk 20]
```diff
@@ -40,59 +41,139 @@ var _ = Suite(&AvahiObserveInterfaceSuite{
...
+			Snap: &snap.Info{
+				SuggestedName: "avahi-observe",
+				Apps:          map[string]*snap.AppInfo{"app": app},
+			},
+			Name:      "avahi-observe",
+			Interface: "avahi-observe",
+			Apps:      map[string]*snap.AppInfo{"app": app},
+		},
+	}
+	release.OnClassic = false
```
Comment (zyga) [inline interfaces/builtin/avahi_observe_test.go]: Please use the right mocking helper here.

File: interfaces/builtin/avahi_observe_test.go [hunk 21]
```diff
@@ -40,59 +41,139 @@ var _ = Suite(&AvahiObserveInterfaceSuite{
...
+	apparmorSpec := &apparmor.Specification{}
+	err := apparmorSpec.AddConnectedPlug(s.iface, s.plug, nil, slot, nil)
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SecurityTags(), DeepEquals, []string{"snap.other.app2"})
+	c.Assert(apparmorSpec.SnippetForTag("snap.other.app2"), testutil.Contains, `peer=(label="snap.avahi-observe.app"),`)
+}
+
+func (s *AvahiObserveInterfaceSuite) TestConnectedPlugSnippedUsesUnconfinedLabelOnClassic(c *C) {
+	slot := &interfaces.Slot{}
+	release.OnClassic = true
```
Comment (zyga) [inline interfaces/builtin/avahi_observe_test.go]: Please use the right mocking helper here.

**Comment by zyga** [comment]
I'll push some cleanups into this PR, please don't close it

**Comment by zyga** [comment]
I merged master and resolved API breaking conflicts. Let's please iterate on this branch to address remaining points and one @jdstrand is happy, land it.

**Comment by zyga** [comment]
I think this is ready for review. @jdstrand please note that the two interfaces, avahi-{observe,control} share some snippets. I would perhaps rather have them explicitly shared (e.g. using a common 3rd snipped) or just copy-paste the relevant parts to the primary snippets. Please advise on how you'd like to see this.

@kubiko I updated the PR and re-worked tests (have a look). I cut tests that were just checking the slot/plug app pattern generator and instead focused on testing where and what kind of snippets do we generate. I also tweaked how tests are prepared, how variables and snap apps are named, all for readability and (hopefully) some more consistency with other interfaces. This is constantly evolving but I'm trying my best to cut the amount of useless code and signal/noise ratio.

File: interfaces/builtin/avahi_control.go [hunk 22]
```diff
@@ -124,25 +124,29 @@ func (iface *avahiControlInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
 	} else {
 		new = slotAppLabelExpr(slot)
 	}
-	snippet := strings.Replace(avahiObserveConnectedPlugAppArmor+avahiControlConnectedPlugAppArmor, old, new, -1)
+	snippet := strings.Replace(avahiControlConnectedPlugAppArmor+avahiControlConnectedPlugAppArmor, old, new, -1)
```
Comment (kubiko) [inline interfaces/builtin/avahi_control.go]: are you sure about this? Idea here is that control also includes observe part.
avahiControlConnectedPlugAppArmor + avahiControlConnectedPlugAppArmor seems odd to me

File: interfaces/builtin/avahi_control.go [hunk 23]
```diff
@@ -124,25 +124,29 @@ func (iface *avahiControlInterface) AppArmorConnectedPlug(spec *apparmor.Specifi
...
+	// base declarations for running as the avahi service.
 	spec.AddSnippet(avahiObservePermanentSlotAppArmor)
 	return nil
 }
 
 func (iface *avahiControlInterface) AppArmorConnectedSlot(spec *apparmor.Specification, plug *interfaces.Plug, plugAttrs map[string]interface{}, slot *interfaces.Slot, slotAttrs map[string]interface{}) error {
 	old := "###PLUG_SECURITY_TAGS###"
 	new := plugAppLabelExpr(plug)
-	snippet := strings.Replace(avahiObserveConnectedSlotAppArmor+avahiControlConnectedSlotAppArmor, old, new, -1)
+	snippet := strings.Replace(avahiControlConnectedSlotAppArmor+avahiControlConnectedSlotAppArmor, old, new, -1)
```
Comment (kubiko) [inline interfaces/builtin/avahi_control.go]: same here

File: interfaces/builtin/avahi_control.go (hunk 22 above)
Comment (zyga) [inline interfaces/builtin/avahi_control.go]: Well, this is *avahi_control* and otherwise the *control* snippet was just unused. Perhaps the rules need tweaking but this was a clear mistake IMO.

File: interfaces/builtin/avahi_control.go (hunk 23 above)
Comment (zyga) [inline interfaces/builtin/avahi_control.go]: Same as above.

File: interfaces/builtin/avahi_control.go [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+	old := "###SLOT_SECURITY_TAGS###"
+	var new string
+	if release.OnClassic {
+		// If we're running on classic Avahi will be part
+		// of the OS snap and will run unconfined.
+		new = "unconfined"
+	} else {
+		new = slotAppLabelExpr(slot)
+	}
+	snippet := strings.Replace(avahiObserveConnectedPlugAppArmor, old, new, -1)
```
Comment (jdstrand) [inline interfaces/builtin/avahi_control.go]: I think it is fine to re-use this variable, but can you make a comment here. I suggest:
```
# avahi-control implies avahi-observe, so add snippets for both here
```

File: interfaces/builtin/avahi_observe.go [hunk 25]
```diff
@@ -19,14 +19,212 @@
...
+  <allow send_destination="org.freedesktop.Avahi"/>
+  <allow receive_sender="org.freedesktop.Avahi"/>
+</policy>
+<policy user="root">
+  <allow send_destination="org.freedesktop.Avahi"/>
+  <allow receive_sender="org.freedesktop.Avahi"/>
+</policy>
 `
 
 const avahiObserveConnectedPlugAppArmor = `
```
Comment (jdstrand) [inline interfaces/builtin/avahi_observe.go]: Same here for avahiObserveConnectedPlugAppArmor:
```
# Note: avahiObserveConnectedPlugAppArmor is also used by avahi-control in AppArmorConnectedPlug
```

File: interfaces/builtin/avahi_control.go [hunk 26]
```diff
@@ -0,0 +1,171 @@
...
+		spec.AddSnippet(avahiObservePermanentSlotAppArmor)
+	}
+	return nil
+}
+
+func (iface *avahiControlInterface) AppArmorConnectedSlot(spec *apparmor.Specification, plug *interfaces.Plug, plugAttrs map[string]interface{}, slot *interfaces.Slot, slotAttrs map[string]interface{}) error {
+	if !release.OnClassic {
+		old := "###PLUG_SECURITY_TAGS###"
+		new := plugAppLabelExpr(plug)
+		snippet := strings.Replace(avahiObserveConnectedSlotAppArmor, old, new, -1)
```
Comment (jdstrand) [inline interfaces/builtin/avahi_control.go]: Same here:
```
# avahi-control implies avahi-observe, so add snippets for both here
```

File: interfaces/builtin/avahi_observe.go [hunk 27]
```diff
@@ -19,14 +19,212 @@
...
+
+# Allow traffic to/from org.freedesktop.DBus for Avahi service
+dbus (receive, send)
+   bus=system
+   path=/org/freedesktop/Avahi{,/**}
+   interface=org.freedesktop.DBus.*
+   peer=(label=unconfined),
+`
+
+const avahiObserveConnectedSlotAppArmor = `
```
Comment (jdstrand) [inline interfaces/builtin/avahi_observe.go]: Also, since avahiObserveConnectedSlotAppArmor is used in avahi-control, let's mention that above avahiObserveConnectedSlotAppArmor:
```
# Note: avahiObserveConnectedSlotAppArmor is also used by avahi-control in AppArmorConnectedSlot
```

File: interfaces/builtin/avahi_observe.go [hunk 28]
```diff
@@ -19,14 +19,212 @@
...
+    interface=org.freedesktop.Avahi.ServiceTypeBrowser
+    peer=(label=###PLUG_SECURITY_TAGS###),
+
+dbus (send)
+    bus=system
+    interface=org.freedesktop.Avahi.ServiceTypeBrowser
+    peer=(name=org.freedesktop.Avahi, label=###PLUG_SECURITY_TAGS###),
+`
+
+const avahiObservePermanentSlotDBus = `
```
Comment (jdstrand) [inline interfaces/builtin/avahi_observe.go]: And here:
```
# Note: avahiObservePermanentSlotDBus is used by avahi-control in DBusPermanentSlot
```

File: interfaces/builtin/avahi_control_test.go [hunk 29]
```diff
@@ -0,0 +1,171 @@
...
+	// on a core system with avahi slot coming from a regular app snap.
+	restore := release.MockOnClassic(false)
+	defer restore()
+
+	// connected plug to app slot
+	spec := &apparmor.Specification{}
+	c.Assert(spec.AddConnectedPlug(s.iface, s.plug, nil, s.appSlot, nil), IsNil)
+	c.Assert(spec.SecurityTags(), DeepEquals, []string{"snap.consumer.app"})
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "name=org.freedesktop.Avahi")
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, `peer=(label="snap.producer.app"),`)
```
Comment (jdstrand) [inline interfaces/builtin/avahi_control_test.go]: It might be nice to verify a specific rule from avahi-observe and from avahi-control for AddConnectedPlug and AddConnectedSlot since avahi-control pulls in both.

File: interfaces/builtin/avahi_observe.go [hunk 30]
```diff
@@ -19,14 +19,212 @@
...
+import (
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/dbus"
+	"github.com/snapcore/snapd/release"
+)
+
 const avahiObserveSummary = `allows discovering local domains, hostnames and services`
```
Comment (kubiko) [inline interfaces/builtin/avahi_observe.go]: Done

File: interfaces/builtin/avahi_observe.go (hunk 27 above)
Comment (kubiko) [inline interfaces/builtin/avahi_observe.go]: Done

File: interfaces/builtin/avahi_observe.go [hunk 31]
```diff
@@ -19,14 +19,212 @@
...
+
+  <deny send_destination="org.freedesktop.Avahi"
+        send_interface="org.freedesktop.Avahi.Server" send_member="SetHostName"/>
+</policy>
+
+<!-- Allow everything, including access to SetHostName to users of the group "netdev" -->
+<policy group="netdev">
+  <allow send_destination="org.freedesktop.Avahi"/>
+  <allow receive_sender="org.freedesktop.Avahi"/>
+</policy>
```
Comment (kubiko) [inline interfaces/builtin/avahi_observe.go]: Good find, indeed makes no sense here.
I removed the policy and added comment why so.

File: interfaces/builtin/avahi_observe.go (hunk 25 above)
Comment (kubiko) [inline interfaces/builtin/avahi_observe.go]: Done

Comment (kubiko) [inline interfaces/builtin/avahi_observe.go]: Done

File: interfaces/builtin/avahi_observe.go [hunk 32]
```diff
@@ -38,14 +236,19 @@ dbus (send)
...
 dbus (send)
     bus=system
     path=/
     interface=org.freedesktop.Avahi.Server
-    member=Get*
-    peer=(name=org.freedesktop.Avahi,label=unconfined),
+    member={Get*,Resolve*,IsNSSSupportAvailable}
+    peer=(name=org.freedesktop.Avahi,label=###SLOT_SECURITY_TAGS###),
+
+dbus (receive)
```
Comment (kubiko) [inline interfaces/builtin/avahi_observe.go]: Done

File: interfaces/builtin/avahi_control.go [hunk 33]
```diff
@@ -0,0 +1,171 @@
...
+        - app
+        - core
+    deny-auto-connection: true
+    deny-connection:
+      on-classic: false
+`
+
+const avahiControlSummary = `allows control over service discovery on a local network via the mDNS/DNS-SD protocol suite`
+
+const avahiControlConnectedSlotAppArmor = `
```
Comment (kubiko) [inline interfaces/builtin/avahi_control.go]: Done

File: interfaces/builtin/avahi_control.go (hunk 24 above)
Comment (kubiko) [inline interfaces/builtin/avahi_control.go]: Done

File: interfaces/builtin/avahi_control.go (hunk 26 above)
Comment (kubiko) [inline interfaces/builtin/avahi_control.go]: Done

File: interfaces/builtin/avahi_observe.go (hunk 28 above)
Comment (kubiko) [inline interfaces/builtin/avahi_observe.go]: Done

File: interfaces/builtin/avahi_observe.go [hunk 34]
```diff
@@ -38,14 +236,19 @@ dbus (send)
...
-    peer=(name=org.freedesktop.Avahi,label=unconfined),
+    peer=(name=org.freedesktop.Avahi,label=###SLOT_SECURITY_TAGS###),
 
 dbus (send)
     bus=system
     path=/
     interface=org.freedesktop.Avahi.Server
-    member=Get*
-    peer=(name=org.freedesktop.Avahi,label=unconfined),
+    member={Get*,Resolve*,IsNSSSupportAvailable}
```
Comment (kubiko) [inline interfaces/builtin/avahi_observe.go]: when I was testing it with third party apps,  some are using instead of {Host/Address/Domain}Resolver direct resolve functions (ResolveHostName, ResolveAddress, ResolveService, ….). Then I was split to either use separate rules and add it under each respective resolving section, or to add it to existing Get* rule all at once. Similar for IsNSSSupportAvailable  which did not fit anywhere.  I did not have strong opinion here, smaller amounts of rules won eventually.
But happy to split them and add each to respective sections, if that improves readability.

File: interfaces/builtin/avahi_control_test.go (hunk 29 above)
Comment (kubiko) [inline interfaces/builtin/avahi_control_test.go]: Tests are now improved for both observe and control to check if observe does not include control and if control includes observe

**Comment by kubiko** [comment]
@jdstrand I now pushed changes in, let me know if you want me to split
`member={Get*,Resolve*,IsNSSSupportAvailable}`
and include each rule in respective sections. Rational behind is explained in comment above, again I have no strong opinion either way, so whatever fits better with other interfaces implementation :)
Rest of the comments should be addressed in commit

**Comment by zyga** [comment]
@kubiko I suggest that you install some plugin for your editor to `go fmt` the code on each save.

```
Formatting wrong in following files:
/tmp/static-unit-tests/src/github.com/snapcore/snapd/interfaces/builtin/avahi_control_test.go
/tmp/static-unit-tests/src/github.com/snapcore/snapd/interfaces/builtin/avahi_observe.go
/tmp/static-unit-tests/src/github.com/snapcore/snapd/interfaces/builtin/avahi_observe_test.go
/tmp/static-unit-tests/src/github.com/snapcore/snapd/interfaces/builtin/avahi_control_test.go
/tmp/static-unit-tests/src/github.com/snapcore/snapd/interfaces/builtin/avahi_observe.go
/tmp/static-unit-tests/src/github.com/snapcore/snapd/interfaces/builtin/avahi_observe_test.go
```

**Comment by kubiko** [comment]
formating now fixed as well

File: interfaces/builtin/avahi_observe.go (hunk 34 above)
Comment (jdstrand) [inline interfaces/builtin/avahi_observe.go]: I think it is fine to leave them combined, but perhaps add this comment:
```
# Allow accessing DBus properties and resolving
```

**Comment by kubiko** [comment]
@jdstrand thanks you
I now added comment you mentioned.

**Comment by niemeyer** [comment]
Great collaboration in this PR. Thanks to all involved.
